src/Microsoft.IdentityModel.Tokens/ProvidersCache/HashCache.cs

```diff
@@ -0,0 +1,264 @@
+        /// <param name="key">The key of the element to remove and return.</param>
+        /// <param name="value">The object removed from the cache, or the default value of the TValue type if key does not exist.</param>
+        /// <returns>true if the object was removed successfully; otherwise, false.</returns>
+        protected virtual bool TryRemoveInternal(TKey key, out ProviderCacheItem<TKey, TValue> value)
+        {
+            value = default;
+
+            var found = _map.TryRemove(key, out ProviderCacheItem<TKey, TValue> val);
+            if (found)
+                value = val;
+
+            return found;
+        }
+
+        /// <summary>
+        /// Validate the key, value and check the expiration time before adding the item is added to the cache.
+        /// </summary>
+        /// <param name="key">The key of the item to be added to the cache.</param>
+        /// <param name="value">The value of the item to be added to the cache.</param>
+        /// <param name="expirationTime">The expiration time of the item.</param>
+        /// <returns>true if valid, false otherwise.</returns>
+        protected static bool ValidateValues(TKey key, TValue value, DateTime expirationTime)
+        {
+            if (key == null)
+                throw LogHelper.LogArgumentNullException(nameof(key));
+
+            if (value == null)
+                throw LogHelper.LogArgumentNullException(nameof(value));
+
+            // if item already expired, do not add it to the cache
+            return expirationTime >= DateTime.UtcNow;
+        }
+
+        /// <summary>
+        /// When the cache is at _maxCapacityPercentage, it needs to be compacted by _compactionPercentage.
+        /// This method calculates the new size of the cache after being compacted.
+        /// </summary>
+        /// <returns>The new target cache size after compaction.</returns>
+        protected int CalculateNewCacheSize()
+        {
+            // use the smaller of _map.Count and _capacity
+            int currentCount = Math.Min(_map.Count, _capacity);
+            return currentCount - (int)(currentCount * _compactionPercentage);
+        }
+
+        /// <summary>
+        /// Determine if the cache needs compaction.
+        /// </summary>
+        protected bool NeedsCompaction => (double)_map.Count / _capacity >= _maxCapacityPercentage;
+
+        #endregion
+
+        /// <summary>
+        /// Remove items from the map by the desired compaction percentage.
+        /// Since this is a simple hash-based cache that does not track the last used time of items, simply remove the desired number of items.
+        /// This should be a private method.
+        /// </summary>
+        private void CompactCache(object state)
+        {
+            // use the _capacity for the newCacheSize calculation in the case where the cache is experiencing overflow
+            var newCacheSize = CalculateNewCacheSize();
+            while (_map.Count > newCacheSize)
```

Comment:
>_map.Count [](http://example.com/codeflow?start=19&length=10)

if a process is filling this up very fast we should have a signal for that situation. E.g. if we never get out of this state the service should know. #operability

Start time for compacting and end time for compacting. total time compacting should likely also be recorded. Average time compacting. Number of times compacting e.g. per hour/day would also be good numbers. 

Tim suggestions here?

Reply:
This while loop is similar to the while loop in the EventBasedLRUCache event queue task, except that it should be faster as the operations are simpler. I have unit tests that covers the cache overflow scenarios with multiple threads and things work well. We can certainly enhance this with signals but that should be in a separate PR.

Reply:
I have created a task to add the measure the new cache in the PerfAndStree, https://identitydivision.visualstudio.com/Engineering/_workitems/edit/1672505/

Reply:
@henrik-me excellent idea, are you blocked on adding these telemetry signals?